add tests for matrix_solver.py

nuclearkatie opened the pull request:
4 (very simple) tests for the matrix solver. closes #27

bcornille commented:
For larger matrices, you can simply start with `A` and `x` then multiply them to get `b`, which can then be solved and checked with the original `x`. Surprisingly, random matrix elements usually still give non-singular and well-conditioned matrices. I recommend having `A` be random and `x` be an vector of `1`s or random. Setting `A` to be only ones will not work.

nuclearkatie commented:
@bcornille and/or @gonuke, can you review how I updated `matrix_solver` to raise singular matrix exceptions, and then the test associated with that update (`test_singular`)? I'm not sure the most pythonic way to do that

bcornille commented on src/matrix_solver.py:
```diff
@@ -12,8 +12,12 @@ def solve_matrix(A, b):
 
-    temperature_array = solve(A, b)
+    try:
+        i = np.linalg.inv(A)
```
This line takes substantially more work than solving the system Ax = b and `i` is not used.

gonuke replied:
Is this meant as a test for singularity of the matrix?  [Here is a Stackoverflow discussion](https://stackoverflow.com/questions/13249108/efficient-pythonic-check-for-singular-matrix) on how to do this.

bcornille commented on tests/test_matrix_solver.py:
```diff
@@ -0,0 +1,59 @@
+
+def test_output_4():
+
+    N = randint(1, 20)
```
Instead of a single test at a random size, consider parameterizing this test to run for several sizes.

bcornille commented on tests/test_matrix_solver.py:
```diff
@@ -0,0 +1,59 @@
+def test_singular():
+
+    N = randint(1, 20)
+    A = np.arange(N**2).reshape((N, N))
```
A matrix of ones is probably more clear that it will be singular.

bcornille commented on tests/test_matrix_solver.py:
```diff
@@ -0,0 +1,59 @@
+
+def test_singular():
+
+    N = randint(1, 20)
```
Same note as `test_output_4`

bcornille commented on src/matrix_solver.py:
```diff
@@ -12,8 +12,12 @@ def solve_matrix(A, b):
-    return temperature_array
+        return temperature_array
+    except np.linalg.LinAlgError as err:
+        return False
```
Instead of returning `False` maybe consider terminating the program and outputting an error. If done this way you will have to wrap every call to this function in something to check for this. Otherwise you may have confused developers wondering why they get a type error when trying to work with the `temperature_array` later on.

gonuke replied:
Agreed - it is not consistent with your interface to return something of a different type than advertised.

gonuke commented on src/matrix_solver.py:
```diff
@@ -12,8 +12,12 @@ def solve_matrix(A, b):
-    temperature_array = solve(A, b)
+    try:
+        i = np.linalg.inv(A)
+        temperature_array = linalg.solve(A, b)
```
you should consistently refer to either `np.linalg` or `linalg`.  I suspect both are valid in this code given the way you have imported, but the difference may be confusing to some.

gonuke commented on tests/test_matrix_solver.py:
```diff
@@ -0,0 +1,59 @@
+    A = np.array([[5, -6, -7], [6, -4, 10], [2, 4, -3]])
+    b = np.array([[7], [-34], [29]])
+
+    assert np.allclose(sm(A, b), np.array([[2], [4], [-3]]))
```
might also consider using `pytest.approx` - a discussion of the relative merits is a the end of [this subsection of pytest documentation](https://docs.pytest.org/en/latest/reference.html#pytest-approx)

nuclearkatie commented:
@bcornille and @gonuke I believe I addressed all your comments
* I took out everything about singular matrices and will just allow `matrix_solver` to raise a LinAlg error if one ever gets passed to the function
* I parametrized `test_output_4` to test a few matrix sizes
* I replaced `np.allclose()` with `pytest.approx()`

nuclearkatie commented:
Any final comments @gonuke? Hoping to get this merged soon
